tofhir-engine/src/main/scala/io/tofhir/engine/data/write/SparkArchiver.scala

```diff
@@ -0,0 +1,179 @@
+    // check if archive mode is enabled for that job
+    mappingJobExecution.job.dataProcessingSettings.archiveMode match {
+      case ArchiveModes.OFF => // do nothing
+      case _ =>
+        mappingJobExecution.job.sourceSettings.head._2 match {
+          case fileSystemSourceSettings: FileSystemSourceSettings => this.applyArchiving(df, mappingJobExecution)
+          case _ => // do nothing, we only support file system source settings for archiving
+        }
+    }
+  }
+
+  /**
+   * Apply archiving for file type data sources e.g. move to archive folder, delete from source folder
+   * @param df
+   * @param mappingJobExecution
+   */
+  private def applyArchiving(df: Dataset[FhirMappingResult], mappingJobExecution: FhirMappingJobExecution): Unit = {
+    val fileSystemSourceSettings = mappingJobExecution.job.sourceSettings.head._2.asInstanceOf[FileSystemSourceSettings]
+    // get data folder path from data source settings
+    val dataFolderPath = fileSystemSourceSettings.dataFolderPath
+    if (fileSystemSourceSettings.asStream) {
+      // get file names from df by grouping by path
+      val schema = schema_of_json(df.collect().head.source.get)
+      // uri paths are included in the dataframes for the streaming jobs, get by grouping by filename
+      val absoluteFilePaths = df
+        .withColumn("jsonData", from_json(col("source"), schema))
+        .select("jsonData.*")
+        .groupBy(col("filename")).count().collect().map(x => x.getString(0))
+      // create path object for each file paths e.g. file:///C:/dev/be/data-integration-suite/test-data/streaming-folder/patients/patients-invalid-input.csv
+      val absolutePaths = absoluteFilePaths.map(x => Paths.get(new URI(x)))
+      // create archive path
+      absolutePaths.foreach(absolutePath => {
+        // create a relative path to the context path of the engine for that file
+        val relativeFilePath = FileUtils.getPath("").toAbsolutePath.relativize(absolutePath)
+        this.processArchiveMode(mappingJobExecution, relativeFilePath)
+      })
+    } else {
+      // find specific file path by matching mapping urls in the job and df
+      df.select("mappingUrl")
+        .groupBy(col("mappingUrl")).count().collect().map(x => x.getString(0))
+        .foreach((mappingUrl: String) => {
+          // find file name by mapping url
+          val paths = mappingJobExecution.job.mappings.find(x => x.mappingRef == mappingUrl).get.sourceContext.map(fhirMappingSourceContextMap => {
+            fhirMappingSourceContextMap._2 match {
+              case fileSystemSource: FileSystemSource => fileSystemSource.path
+            }
+          })
+          paths.foreach(relativePath => {
+            val relativeFilePath = Paths.get(dataFolderPath, relativePath)
+            this.processArchiveMode(mappingJobExecution, relativeFilePath)
+          })
+        })
+    }
+  }
+
+  /**
+   * Process archive mode according to the job settings
```

Comment:
Can you please extend the description according to the archiving options available?